### PR TITLE
Convert some of appsupport.js functions to Rescript

### PR DESCRIPTION
**Pull request by OceanOak**
Changelog:

```
Area of change
- details
```
**Questions:**
- Do we want to create a new file appsupport.res or do we want to move the functions somewhere else?
- Still not sure how to test these functions.

### CODE REVIEW

File: client/src/appsupport.res [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+    let reload = ref(false)
+    let observer= MutationObserver.make((_records, _observer) => ())
+    let records: array<Dom.mutationRecord> = observer->MutationObserver.takeRecords
+    records|>Js.Array.forEach(record=>{
+      record->removedNodes
+      |> Webapi.Dom.NodeList.toArray
+      |> Js.Array.forEach(node=>{
+        Js.log(node)
+        // if (node["id"]== "unsupportedBrowser") {
+        //   reload := true
```
Comment (OceanOak) [inline client/src/appsupport.res:55]: I couldn't find node.id in rescript-webapi

File: client/src/appsupport.res [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+  open! KeyboardEvent
+  // Don't ever attempt to save the HTML of the page.
+ if (ctrlKey(event)|| metaKey(event)) && key(event)=="s"{
+  event -> preventDefault
+ }
+
+ // `Ctrl-K` is meant to open the omnibox on Linux, but without this preventDefault
+ // it will focus the browser's URL bar after creating the omnibox.
+
+  if ctrlKey(event)&& key(event)=="k" && Entry.getBrowserPlatform() == Linux{
```
Comment (OceanOak) [inline client/src/appsupport.res:98]: `window.navigator.platform `isn't found in neither webapi nor rescript's api.
`navigator.userAgentData.platform` could be an alternative but it isn't found in webapi as well.
I've used `Entry.getBrowserPlatform` but I'm not sure if it is the right choice.

File: client/src/appsupport.res [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+  }
+  let configure = ({enabled: enabled, payload: payload}) => {
+    rollbar.configure({enabled: enabled, payload: payload})
+  }
+
+let username=""
+let userID=""
+  let rollbarConfig = {
+    enabled: true,
+    payload: { id: userID, username: username }
```
Comment (OceanOak) [inline client/src/appsupport.res:168]: I am not sure where to get username and userID from. Used empty strings just to pass this step.

File: client/src/appsupport.res [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+let pageHidden = ref(false)
+
+let visibilityCheck=() => {
+  let hidden = ref(false)
+  if (Js.typeof(Webapi.Dom.document |> Document.hidden) !== "undefined") {
+    hidden := Webapi.Dom.document |> Document.hidden
+  }
+  // else if (typeof document.mozHidden !== "undefined") {
+  //   hidden = document.mozHidden
+  // } else if (typeof document.msHidden !== "undefined") {
```
Comment (OceanOak) [inline client/src/appsupport.res:252]: Couldn't find `document.mozHidden` in rescript-webapi.

File: client/src/appsupport.res (hunk 2 above)
Comment (pbiggar) [reply client/src/appsupport.res:98]: Definitely worth adding it to rescript-webapi if it's a standard feature that's missing. Then you can use the same definitions in our codebase until we can use the webapi ones.

File: client/src/appsupport.res (hunk 3 above)
Comment (pbiggar) [reply client/src/appsupport.res:168]: Those are created in UI.html: https://github.com/darklang/dark/blob/main/backend/src/ApiServer/Templates/ui.html#L106

File: client/src/appsupport.res (hunk 4 above)
Comment (pbiggar) [reply client/src/appsupport.res:252]: You can add them to webapi and use local versions of the definition here too.

File: client/src/appsupport.res (hunk 3 above)
Comment (StachuDotNet) [reply client/src/appsupport.res:168]: > I am not sure where to get username and userID from.

based on Paul's link, they should just 'be available' to you in `appsupport.js` - the values are 'injected' when you load the page, from `backend/src/ApiServer/Ui.fs` into `ui.html`, such that both `username` and `userID` should be global variables. If you're not seeing values for them, something may be wrong.

Comment (OceanOak) [reply client/src/appsupport.res:168]: They are available in `appsupport.js`. I need them in `appsupport.res` (the Rescript version of appsupport.js). 
Still not sure if I am supposed to create this file (appsupport.res) or if I should move the functions somewhere else.

Comment (StachuDotNet) [reply client/src/appsupport.res:168]: Look at client/workers/Fetcher.res - I think you can use the pattern of

`@val external buildHash: string = "buildHash"`

to read those globals and pull them into ReScript world?

Comment (StachuDotNet) [reply client/src/appsupport.res:168]: e.g. add  `@val external username: string = "username"` to the top of `appsupport.res`, maybe?

Comment (StachuDotNet) [reply client/src/appsupport.res:168]: > Still not sure if I am supposed to create this file (appsupport.res) or if I should move the functions somewhere else.
I'm not sure (depends on the fns) but moving into an appsupport.res seems a great step forward. Pulling the fns into other files can be in a follow-up PR, I think.

Comment (OceanOak) [reply client/src/appsupport.res:168]: > e.g. add `@val external username: string = "username"` to the top of `appsupport.res`, maybe?

This worked and will enable me to convert more functions. Thank you so much! (I still need to test all of these functions to ensure they work properly.)

**Comment by pbiggar** [comment]
Client is gone

**Pull request closed without merging**